src/workerd/api/streams/writable.c++

```diff
@@ -369,6 +369,173 @@ private:
+      : context(context), writer(kj::mv(writer)) {}
+
+  ~WritableStreamJsRpcAdapter() noexcept(false) {
+    weakRef->invalidate();
+    doneFulfiller->fulfill();
+
+    // If the stream was not explicitly ended and the writer still exists at this point,
+    // then we should trigger calling the abort algorithm on the stream. Sadly, there's a
+    // bit of an incompetibility with kj::AsyncOutputStream and the standard definition of
+    // WritableStream in that AsyncOutputStream has no specific way to explicitly signal that
+    // the stream is being aborted, or even being closed due to a particular reason. On the
+    // remote side, because it is using a WritableStreamSink implementation, when that side
+    // is aborted, all it does is record the reason and drop the stream. It does not propagate
+    // the reason back to this side. So, we have to do the best we can here. Our assumption
+    // is that once the stream is dropped, if it has not been explicitly ended and the writer
+    // still exists, then the writer should be aborted. This is not perfect because we
+    // cannot propgate the actual reason why it was aborted.
+    //
+    // Note also that there is no guarantee that the abort will actually run if the context
+    // is being torn down. Some WritableStream implementations might use the abort algorithm
+    // to clean things up or perform logging in the case of an error. Care needs to be taken
+    // in this situation or the user code might end up with bugs. Need to see if there's a
+    // better solution.
+    //
+    // TODO(someday): If the remote end can be updated to propagate the abort, then we can
+    // hopefully improve the situation here.
+    if (!ended) {
+      KJ_IF_SOME(writer, this->writer) {
+        context.addTask(context.run(
+            [writer=kj::mv(writer), exception=cancellationException()]
+            (Worker::Lock& lock) mutable {
+          jsg::Lock& js = lock;
+          auto ex = js.exceptionToJs(kj::mv(exception));
+          return IoContext::current().awaitJs(lock,
+              writer->abort(lock, ex.getHandle(js)));
+        }));
+      }
+    }
+  }
+
+  // Returns a promise that resolves when the stream is dropped. If the promise is canceled before
+  // that, the stream is revoked.
+  kj::Promise<void> waitForCompletionOrRevoke() {
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    doneFulfiller = kj::mv(paf.fulfiller);
+
+    return paf.promise.attach(kj::defer([weakRef = weakRef->addRef()]() mutable {
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(cancellationException());
+        }
+      }
+    }));
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    if (size == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run([this, buffer, size](Worker::Lock& lock) mutable {
```

Comment:
For review: note that this approach does mean that we end up having to reacquire the lock on *every write*. Which is painful. Another approach we could take here is to push the writes onto a buffer and separately fire up the `context.run(...)` such that once the lock is acquired that buffer is drained. We could potentially coalesce multiple writes within a single lock. Went with the less sophisticated approach here for now but likely a good idea to optimize here as a next step.

Reply:
We should probably optimize `context.run` itself so that it coalesces jobs rather than repeatedly re-acquiring the lock. (Not in this PR.)